File: 20230205_Sunday/20230205.go
// Leetcode: 438. Find All Anagrams in a String.
// https://leetcode.com/problems/find-all-anagrams-in-a-string/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 4 ms, faster than 91.22% of Go online submissions for Find All Anagrams in a String.
// Memory Usage: 5 MB, less than 65.16% of Go online submissions for Find All Anagrams in a String.
// 2023.02.05 Daily Challenge.
func findAnagrams(s string, p string) []int {
  sl,pl := len(s), len(p)
  if sl < pl {
    return make([]int,0)
  }
  tmplt, wnd := count(p),count(s[0:pl])
  answer := make([]int,0)
  if same(tmplt, wnd) {
    answer = append(answer,0)
  }
  // i - index to decrease.
  // j - index to insert in answer.
  // k - index to increase in counter. 
  for i,j,k := 0, 1, pl; k < sl; i,j,k = i+1, j+1, k+1 {
    wnd[s[i] - 'a'] -= 1
    wnd[s[k] - 'a'] += 1
    if same(wnd,tmplt) {
      answer = append(answer,j)
    }
  }
  return answer
}
func count(str string) []int{
  answer := make([]int,26)
  for _,c := range str {
    answer[c - 'a'] += 1
  }
  return answer
}
func same(a,b []int) bool{
  for i,v := range a {
    if v != b[i] { 
      return false
    }
  }
  return true
}

File: 20230205_Sunday/20230205.rb
# Leetcode: 438. Find All Anagrams in a String.
# https://leetcode.com/problems/find-all-anagrams-in-a-string/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 202 ms, faster than 92.16% of Ruby online submissions for Find All Anagrams in a String.
# Memory Usage: 212.4 MB, less than 76.47% of Ruby online submissions for Find All Anagrams in a String.
# 2023.02.05 Daily Challenge.
# @param {String} s
# @param {String} pp
# @return {Integer[]}
def find_anagrams(s, pp)
  tmpl = get_count(pp) # Template.
  cmpr = get_count(s[0...pp.size]) # For compare.
  answer = []
  answer.push(0) if tmpl == cmpr
  (pp.size...s.size).each do |i|
    prev = s[i-pp.size].ord - ?a.ord
    curr = s[i].ord - ?a.ord
    cmpr[prev] -= 1
    cmpr[curr] += 1
    answer.push(i-pp.size+1) if tmpl == cmpr
  end
  answer
end
def get_count(str)
  answer = [0]*26
  str.chars.each do |c|
    answer[c.ord - ?a.ord] += 1
  end
  return answer
end
